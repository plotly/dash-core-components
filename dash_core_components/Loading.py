# AUTO GENERATED FILE - DO NOT EDIT

from dash.development.base_component import Component, _explicitize_args


class Loading(Component):
    """A Loading component.
A Loading component that wraps any other component and displays a spinner until the wrapped component has rendered.

Keyword arguments:
- children (list of a list of or a singular dash component, string or numbers | a list of or a singular dash component, string or number; optional): Array that holds components to render
- id (string; optional): The ID of this component, used to identify dash components
in callbacks. The ID needs to be unique across all of the
components in an app.
- type (a value equal to: 'graph', 'cube', 'circle', 'dot', 'default'; default 'default'): Property that determines which spinner to show
one of 'graph', 'cube', 'circle', 'dot', or 'default'.
- fullscreen (boolean; optional): Boolean that makes the spinner display full-screen
- debug (boolean; optional): If true, the spinner will display the component_name and prop_name
while loading
- className (string; optional): Additional CSS class for the spinner root DOM node
- parent_className (string; optional): Additional CSS class for the outermost dcc.Loading parent div DOM node
- style (dict; optional): Additional CSS styling for the spinner root DOM node
- parent_style (dict; optional): Additional CSS styling for the outermost dcc.Loading parent div DOM node
- color (string; default '#119DFF'): Primary colour used for the loading spinners
- loading_state (dict; optional): Object that holds the loading state object coming from dash-renderer. loading_state has the following type: dict containing keys 'is_loading', 'prop_name', 'component_name'.
Those keys have the following types:
  - is_loading (boolean; optional): Determines if the component is loading or not
  - prop_name (string; optional): Holds which property is loading
  - component_name (string; optional): Holds the name of the component that is loading"""
    @_explicitize_args
    def __init__(self, children=None, id=Component.UNDEFINED, type=Component.UNDEFINED, fullscreen=Component.UNDEFINED, debug=Component.UNDEFINED, className=Component.UNDEFINED, parent_className=Component.UNDEFINED, style=Component.UNDEFINED, parent_style=Component.UNDEFINED, color=Component.UNDEFINED, loading_state=Component.UNDEFINED, **kwargs):
        self._prop_names = ['children', 'id', 'type', 'fullscreen', 'debug', 'className', 'parent_className', 'style', 'parent_style', 'color', 'loading_state']
        self._type = 'Loading'
        self._namespace = 'dash_core_components'
        self._valid_wildcard_attributes =            []
        self.available_properties = ['children', 'id', 'type', 'fullscreen', 'debug', 'className', 'parent_className', 'style', 'parent_style', 'color', 'loading_state']
        self.available_wildcard_properties =            []

        _explicit_args = kwargs.pop('_explicit_args')
        _locals = locals()
        _locals.update(kwargs)  # For wildcard attrs
        args = {k: _locals[k] for k in _explicit_args if k != 'children'}

        for k in []:
            if k not in args:
                raise TypeError(
                    'Required argument `' + k + '` was not specified.')
        super(Loading, self).__init__(children=children, **args)
