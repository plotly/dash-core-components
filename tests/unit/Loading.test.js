import Loading from '../../src/components/Loading.react.js';
import React from 'react';
import {render} from 'enzyme';

test('Loading renders', () => {
    const statusMock = {
        is_loading: true,
        prop_name: 'children',
        component_name: 'div',
    };
    const loading = render(
        <Loading loading_state={statusMock}>
            <div>Loading is done!</div>
        </Loading>
    );

    expect(
        loading
            .find('.dash-spinner')
            .parent()
            .html()
    ).toMatchSnapshot('Loading with is_loading=true');
});
test('Loading renders without loading_state', () => {
    const loading = render(
        <Loading>
            <div>Loading is done!</div>
        </Loading>
    );

    expect(loading.html()).toEqual('<div>Loading is done!</div><div></div>');
});
test('Loading renders without loading_state.is_loading', () => {
    const statusMock = {
        prop_name: 'children',
        component_name: 'div',
    };
    const loading = render(
        <Loading loading_state={statusMock}>
            <div>Loading is done!</div>
        </Loading>
    );

    expect(loading.html()).toEqual('<div>Loading is done!</div><div></div>');
});
test('Loading renders without prop_name', () => {
    const statusMock = {
        is_loading: true,
        component_name: 'div',
    };
    const loading = render(
        <Loading loading_state={statusMock}>
            <div>Loading is done!</div>
        </Loading>
    );

    expect(
        loading
            .find('.dash-spinner')
            .parent()
            .html()
    ).toMatchSnapshot('Loading with is_loading=true');
});
test('Loading renders without loading_state.component_name', () => {
    const statusMock = {
        is_loading: true,
        prop_name: 'children',
    };
    const loading = render(
        <Loading loading_state={statusMock}>
            <div>Loading is done!</div>
        </Loading>
    );

    expect(
        loading
            .find('.dash-spinner')
            .parent()
            .html()
    ).toMatchSnapshot('Loading with is_loading=true');
});
test('Loading renders with multiple children', () => {
    const statusMock = {
        is_loading: true,
        prop_name: 'children',
        component_name: 'div',
    };
    const loading = render(
        <Loading loading_state={statusMock}>
            <div>Child 1</div>
            <div>Child 2</div>
            <div>Child 3</div>
        </Loading>
    );

    expect(
        loading
            .find('.dash-spinner')
            .parent()
            .html()
    ).toMatchSnapshot('Loading with is_loading=true');
});

test("Loading checks all it's children for a loading_state", () => {
    const statusMock = {
        is_loading: true,
        prop_name: 'children',
        component_name: 'div',
    };
    const loading = render(
        <Loading>
            <div>Child 1</div>
            <div>Child 2</div>
            <div loading_state={statusMock}>Child 3</div>
        </Loading>
    );

    expect(loading.html()).toMatchSnapshot('Loading spinner for children');
});
