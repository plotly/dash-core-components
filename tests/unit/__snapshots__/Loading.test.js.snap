// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading checks all it's children for a loading_state: Loading spinner for children 1`] = `"<div>Child 1</div><div>Child 2</div><div loading_state=\\"[object Object]\\">Child 3</div><div></div>"`;

exports[`Loading renders with multiple children: Loading with is_loading=true 1`] = `
"<div class=\\"dash-spinner dash-default-spinner\\"><div class=\\"dash-default-spinner-rect1\\"></div><div class=\\"dash-default-spinner-rect2\\"></div><div class=\\"dash-default-spinner-rect3\\"></div><div class=\\"dash-default-spinner-rect4\\"></div><div class=\\"dash-default-spinner-rect5\\"></div></div><style>
                    .dash-spinner-container {
                        position: fixed;
                        width: 100vw;
                        height: 100vh;
                        top: 0;
                        left: 0;
                        background-color: white;
                        z-index: 99;
                        display: flex;
                        justify-content: center;
                        align-items: center;
                    }
                    .dash-loading-title {
                        text-align: center;
                    }
                    .dash-default-spinner{
                        margin: 1rem auto;
                        width: 50px;
                        height: 40px;
                        text-align: center;
                        font-size: 10px;
                    }

                    .dash-default-spinner &gt; div {
                        background-color: #119DFF;
                        height: 100%;
                        width: 6px;
                        display: inline-block;
                        margin-right: 4px;

                        -webkit-animation: sk-stretchdelay 1.2s infinite ease-in-out;
                        animation: sk-stretchdelay 1.2s infinite ease-in-out;
                    }

                    .dash-default-spinner .dash-default-spinner-rect2 {
                        -webkit-animation-delay: -1.1s;
                        animation-delay: -1.1s;
                    }

                    .dash-default-spinner .dash-default-spinner-rect3 {
                        -webkit-animation-delay: -1.0s;
                        animation-delay: -1.0s;
                    }

                    .dash-default-spinner .dash-default-spinner-rect4 {
                        -webkit-animation-delay: -0.9s;
                        animation-delay: -0.9s;
                    }

                    .dash-default-spinner .dash-default-spinner-rect5 {
                        -webkit-animation-delay: -0.8s;
                        animation-delay: -0.8s;
                    }

                    @-webkit-keyframes sk-stretchdelay {
                        0%, 40%, 100% { -webkit-transform: scaleY(0.4) }
                        20% { -webkit-transform: scaleY(1.0) }
                    }

                    @keyframes sk-stretchdelay {
                        0%, 40%, 100% {
                        transform: scaleY(0.4);
                        -webkit-transform: scaleY(0.4);
                        }  20% {
                        transform: scaleY(1.0);
                        -webkit-transform: scaleY(1.0);
                        }
                    }
            </style>"
`;

exports[`Loading renders without loading_state.component_name: Loading with is_loading=true 1`] = `
"<div class=\\"dash-spinner dash-default-spinner\\"><div class=\\"dash-default-spinner-rect1\\"></div><div class=\\"dash-default-spinner-rect2\\"></div><div class=\\"dash-default-spinner-rect3\\"></div><div class=\\"dash-default-spinner-rect4\\"></div><div class=\\"dash-default-spinner-rect5\\"></div></div><style>
                    .dash-spinner-container {
                        position: fixed;
                        width: 100vw;
                        height: 100vh;
                        top: 0;
                        left: 0;
                        background-color: white;
                        z-index: 99;
                        display: flex;
                        justify-content: center;
                        align-items: center;
                    }
                    .dash-loading-title {
                        text-align: center;
                    }
                    .dash-default-spinner{
                        margin: 1rem auto;
                        width: 50px;
                        height: 40px;
                        text-align: center;
                        font-size: 10px;
                    }

                    .dash-default-spinner &gt; div {
                        background-color: #119DFF;
                        height: 100%;
                        width: 6px;
                        display: inline-block;
                        margin-right: 4px;

                        -webkit-animation: sk-stretchdelay 1.2s infinite ease-in-out;
                        animation: sk-stretchdelay 1.2s infinite ease-in-out;
                    }

                    .dash-default-spinner .dash-default-spinner-rect2 {
                        -webkit-animation-delay: -1.1s;
                        animation-delay: -1.1s;
                    }

                    .dash-default-spinner .dash-default-spinner-rect3 {
                        -webkit-animation-delay: -1.0s;
                        animation-delay: -1.0s;
                    }

                    .dash-default-spinner .dash-default-spinner-rect4 {
                        -webkit-animation-delay: -0.9s;
                        animation-delay: -0.9s;
                    }

                    .dash-default-spinner .dash-default-spinner-rect5 {
                        -webkit-animation-delay: -0.8s;
                        animation-delay: -0.8s;
                    }

                    @-webkit-keyframes sk-stretchdelay {
                        0%, 40%, 100% { -webkit-transform: scaleY(0.4) }
                        20% { -webkit-transform: scaleY(1.0) }
                    }

                    @keyframes sk-stretchdelay {
                        0%, 40%, 100% {
                        transform: scaleY(0.4);
                        -webkit-transform: scaleY(0.4);
                        }  20% {
                        transform: scaleY(1.0);
                        -webkit-transform: scaleY(1.0);
                        }
                    }
            </style>"
`;

exports[`Loading renders without prop_name: Loading with is_loading=true 1`] = `
"<div class=\\"dash-spinner dash-default-spinner\\"><div class=\\"dash-default-spinner-rect1\\"></div><div class=\\"dash-default-spinner-rect2\\"></div><div class=\\"dash-default-spinner-rect3\\"></div><div class=\\"dash-default-spinner-rect4\\"></div><div class=\\"dash-default-spinner-rect5\\"></div></div><style>
                    .dash-spinner-container {
                        position: fixed;
                        width: 100vw;
                        height: 100vh;
                        top: 0;
                        left: 0;
                        background-color: white;
                        z-index: 99;
                        display: flex;
                        justify-content: center;
                        align-items: center;
                    }
                    .dash-loading-title {
                        text-align: center;
                    }
                    .dash-default-spinner{
                        margin: 1rem auto;
                        width: 50px;
                        height: 40px;
                        text-align: center;
                        font-size: 10px;
                    }

                    .dash-default-spinner &gt; div {
                        background-color: #119DFF;
                        height: 100%;
                        width: 6px;
                        display: inline-block;
                        margin-right: 4px;

                        -webkit-animation: sk-stretchdelay 1.2s infinite ease-in-out;
                        animation: sk-stretchdelay 1.2s infinite ease-in-out;
                    }

                    .dash-default-spinner .dash-default-spinner-rect2 {
                        -webkit-animation-delay: -1.1s;
                        animation-delay: -1.1s;
                    }

                    .dash-default-spinner .dash-default-spinner-rect3 {
                        -webkit-animation-delay: -1.0s;
                        animation-delay: -1.0s;
                    }

                    .dash-default-spinner .dash-default-spinner-rect4 {
                        -webkit-animation-delay: -0.9s;
                        animation-delay: -0.9s;
                    }

                    .dash-default-spinner .dash-default-spinner-rect5 {
                        -webkit-animation-delay: -0.8s;
                        animation-delay: -0.8s;
                    }

                    @-webkit-keyframes sk-stretchdelay {
                        0%, 40%, 100% { -webkit-transform: scaleY(0.4) }
                        20% { -webkit-transform: scaleY(1.0) }
                    }

                    @keyframes sk-stretchdelay {
                        0%, 40%, 100% {
                        transform: scaleY(0.4);
                        -webkit-transform: scaleY(0.4);
                        }  20% {
                        transform: scaleY(1.0);
                        -webkit-transform: scaleY(1.0);
                        }
                    }
            </style>"
`;

exports[`Loading renders: Loading with is_loading=true 1`] = `
"<div class=\\"dash-spinner dash-default-spinner\\"><div class=\\"dash-default-spinner-rect1\\"></div><div class=\\"dash-default-spinner-rect2\\"></div><div class=\\"dash-default-spinner-rect3\\"></div><div class=\\"dash-default-spinner-rect4\\"></div><div class=\\"dash-default-spinner-rect5\\"></div></div><style>
                    .dash-spinner-container {
                        position: fixed;
                        width: 100vw;
                        height: 100vh;
                        top: 0;
                        left: 0;
                        background-color: white;
                        z-index: 99;
                        display: flex;
                        justify-content: center;
                        align-items: center;
                    }
                    .dash-loading-title {
                        text-align: center;
                    }
                    .dash-default-spinner{
                        margin: 1rem auto;
                        width: 50px;
                        height: 40px;
                        text-align: center;
                        font-size: 10px;
                    }

                    .dash-default-spinner &gt; div {
                        background-color: #119DFF;
                        height: 100%;
                        width: 6px;
                        display: inline-block;
                        margin-right: 4px;

                        -webkit-animation: sk-stretchdelay 1.2s infinite ease-in-out;
                        animation: sk-stretchdelay 1.2s infinite ease-in-out;
                    }

                    .dash-default-spinner .dash-default-spinner-rect2 {
                        -webkit-animation-delay: -1.1s;
                        animation-delay: -1.1s;
                    }

                    .dash-default-spinner .dash-default-spinner-rect3 {
                        -webkit-animation-delay: -1.0s;
                        animation-delay: -1.0s;
                    }

                    .dash-default-spinner .dash-default-spinner-rect4 {
                        -webkit-animation-delay: -0.9s;
                        animation-delay: -0.9s;
                    }

                    .dash-default-spinner .dash-default-spinner-rect5 {
                        -webkit-animation-delay: -0.8s;
                        animation-delay: -0.8s;
                    }

                    @-webkit-keyframes sk-stretchdelay {
                        0%, 40%, 100% { -webkit-transform: scaleY(0.4) }
                        20% { -webkit-transform: scaleY(1.0) }
                    }

                    @keyframes sk-stretchdelay {
                        0%, 40%, 100% {
                        transform: scaleY(0.4);
                        -webkit-transform: scaleY(0.4);
                        }  20% {
                        transform: scaleY(1.0);
                        -webkit-transform: scaleY(1.0);
                        }
                    }
            </style>"
`;
